### ISSUE TITLE
split template: add top/bottom orientation

### ISSUE THREAD

**Issue by peterpeterparker** [deckgo/deckdeckgo#372]
## Description

The "split" template (https://docs.deckdeckgo.com/slides/split), display currently the content in two columns "left/right".

The idea would be to add a new options in order to be able to display the content in two rows "top/bottom"

This enhancement should find place in the related slide web component.

Once implemented, the option could be added our editor ("studio").

## Questions

1. About the implementation in the editor, for simplicity reason, should we add the "bottom/top" as an option of the split template (new button for option which link to a right popover with the option) or should we "just" add two different templates style in the list of selectable template ("when user add a new slide")

2. Should we go one step further and add more options to the templates for example to be able to display "left/right" on desktop/projector and for the same content "top/bottom" on mobile/tablet. Maybe, to ease the development, if this is cool, we could do that later in another issue?

**Comment by timrodz** [comment]
Heya! I'd be happy to take on this issue. Just found about DeckDeckGo today and I must say it looks great! The docs are superb too. 🕺

**Comment by peterpeterparker** [comment]
Awesome @timrodz and happy to here the docs is cool 👍 

It would be super neat if you could give a hand.

You could find the source of the "split slide/template" in https://github.com/deckgo/deckdeckgo/tree/master/webcomponents/slides/split

```
git clone your-fork-of-deckdeckgo
cd deckdeckgo/webcomponents/slides/split
npm install
npm run start
```

http://localhost:3333/

Not the most really beautiful test page though ;)

Ping me if you have any questions or let me know how I could help.

**Comment by timrodz** [comment]
Thank you, @peterpeterparker! I'll have a go at it 😄

**Comment by timrodz** [comment]
Hi @peterpeterparker, I've had a look at the codebase and had a few questions / thoughts I'd like to verify - I haven't used Stencil before, so I'm quite unfamiliar with the codebase or the way Stencil works.

- According to the docs, every `webcomponent` works as separate npm packages, correct?
- I notice the use of namespacing for specifying settings such as `LocalJSX`, which I assume is the code that developers will inject to match their own implementations.
- For adding the top/bottom split, I'm thinking of creating either:
  - A setting so the user can specify 'vertical'/'horizontal' and it'll render different HTML/CSS | Seems like adding the new split code to [deckdeckgo-slide-split.scss](https://github.com/deckgo/deckdeckgo/blob/master/webcomponents/slides/split/src/components/slide/deckdeckgo-slide-split.scss) would be good enough
  - A new interface along `HTMLDeckgoSlideSplitElement` that will have the changes for the vertical split

**Comment by peterpeterparker** [comment]
* Yes

* I don't understand your question regarding "namespacing and LocalJSX". Could you point out a line of code you are referencing?

* I think a setting respectively a new `@Prop()` is the way. Then, according the value, we could apply a `class` to the container in the `render()` method and finally implement the different display in the `scss` file

**Comment by timrodz** [comment]
Thanks for your answers!

>    I don't understand your question regarding "namespacing and LocalJSX". Could you point out a line of code you are referencing?

Let me rephrase my question: What's the relationship between `components.d.ts` and `deckdeckgo-slide-split.tsx`? Similarly, how do these files talk to each other (assuming they do)

If I take the boolean `customBackground` from `components.d.ts`, I can notice it's referenced inside `Components::DeckgoSlideAuthor` and `LocalJSX::DeckgoSlideAuthor` of the same script.

<img width="707" alt="Screen Shot 2019-10-10 at 9 50 33 PM" src="https://user-images.githubusercontent.com/7434353/66553795-4a97b480-eba8-11e9-8a55-30e74805e620.png">

<img width="573" alt="Screen Shot 2019-10-10 at 9 50 47 PM" src="https://user-images.githubusercontent.com/7434353/66553816-56837680-eba8-11e9-9778-b762b472dfeb.png">

It is also referenced as a `@Prop()` inside `deckdeckgo-slide-split.tsx`, so I'm assuming to define a new prop I'd have to follow similar steps.

<img width="628" alt="Screen Shot 2019-10-10 at 9 53 36 PM" src="https://user-images.githubusercontent.com/7434353/66553876-70bd5480-eba8-11e9-9936-ca9469c73c3a.png">


>   I think a setting respectively a new `@Prop()` is the way (your first option, agree with you). Then, according the value, we could apply a `class` to the container in the `render()` method and finally implement the different display in the `scss` file

Coming from React, I can see accessing variables inside Render() like so:

<img width="582" alt="Screen Shot 2019-10-10 at 9 55 23 PM" src="https://user-images.githubusercontent.com/7434353/66554082-c98ced00-eba8-11e9-929c-43a2f3bb2a09.png">

Based on [Stencil's docs](https://stenciljs.com/docs/properties), I can infer that I can simply use JSX to change the class names and then I'll go along and toy around with `scss` until I get the desired results.

**Comment by peterpeterparker** [comment]
`components.d.ts` is the component's definition file, you could safely ignores it as it is automatically generated by Stencil at build time (`npm run start` or `npm run build`).

Exactly you could define a new`@Prop()` as the other. We don't necessary need the `reflectToAttr` right now, as you wish.

> Based on Stencil's docs, I can infer that I can simply use JSX to change the class names and then I'll go along and toy around with scss until I get the desired results.

Exactly, you read my mind 👍 

P.S.: Note in Stencil JSX class is class, not className as in React

**Comment by timrodz** [comment]
Awesome, thank you! I'll have a jam during the weekend - It's almost there! 🕺

**Comment by peterpeterparker** [comment]
Awesome! Ping me anytime if you've got questions or need help.

**Comment by timrodz** [comment]
Getting there! Just need to add diff properties to the .scss file

<img width="345" alt="Screen Shot 2019-10-12 at 9 46 11 AM" src="https://user-images.githubusercontent.com/7434353/66683906-2685c180-ecd5-11e9-98b2-89b4a8381e89.png">

My approach has been to make it so that we touch as little code as possible:

```tsx
@Prop() isVertical: boolean = false;

render() {
  const verticalAttr = this.isVertical ? '-vertical' : '';
  return <Host class={{'deckgo-slide-container': true}}>
    <div class={`deckgo-slide${verticalAttr}`}>
      <slot name="title"></slot>
      <div class="deckgo-slide-split deckgo-slide-split-start"><slot name="start"></slot></div>
      <div class="deckgo-slide-split deckgo-slide-split-end"><slot name="end"></slot></div>
      <slot name="notes"></slot>
      <slot name="actions"></slot>
      <slot name="background"></slot>
    </div>
  </Host>;
}
```

Looks like I'll probably have to add `verticalAttr` to an alternative for `deckgo-slide-split` so that it can properly calculate the height - What do you think?

**Comment by timrodz** [comment]
Current implementation:

![deckgo_001](https://user-images.githubusercontent.com/7434353/66685867-26d48b80-ecda-11e9-9d9e-aeaaa075ac3c.gif)

It's able to split 50/50 vertically - But I need to take the paddings in the formula, so might have to re-calculate `height` and `width` accordingly:

```scss
div.deckgo-slide-vertical {
  @extend div, .deckgo-slide;
  flex-flow: column wrap;
}

div.deckgo-slide-split-vertical {
  @extend div, .deckgo-slide-split;
  // TODO
}
```

<img width="752" alt="Screen Shot 2019-10-12 at 10 24 06 AM" src="https://user-images.githubusercontent.com/7434353/66686127-e3c6e800-ecda-11e9-9226-000ab2be48c6.png">

**Comment by timrodz** [comment]
Added colours to show the differences between both splits - If you think this is fine, I'll make a Pull Request and we can discuss the code further.

### Horizontal (Default)
<img width="300" src="https://user-images.githubusercontent.com/7434353/66689784-ec261f80-ece8-11e9-90c8-ea5b8e52e550.png"/>

### Vertical
<img width="300" src="https://user-images.githubusercontent.com/7434353/66689785-ec261f80-ece8-11e9-894b-b36b2321e47f.png"/>

**Comment by peterpeterparker** [comment]
It looks good 👍  thx!

Furthermore you teach me something, I never used `@extend` before, thx :)

about size, yes the size is given by the class `.deckgo-slide` which is defined in the util class `@deckdeckgo/slide-utils` (https://github.com/deckgo/deckdeckgo/blob/26a9fb4ec7923cba003296a6c85d95bf245d45e8/webcomponents/slides/utils/styles/deckdeckgo-slide.scss#L23)

Super cool, go for the PR 🚀

**Comment by peterpeterparker** [comment]
Thx you for the neat PR @timrodz 🙏

**Comment by peterpeterparker** [comment]
- [Doc](https://docs.deckdeckgo.com/slides/split/) updated
- Feature unleaded in the [starter](https://github.com/deckgo/deckdeckgo-starter) kit
- Not yet implemented in the editor, see #408
